c10/cuda/CUDAStream.cpp

```diff
@@ -71,6 +70,12 @@ static std::array<LeakyStreamInternals, kStreamsPerPool>
 static std::array<LeakyStreamInternals, kStreamsPerPool>
     high_priority_streams[C10_COMPILE_TIME_MAX_GPUS];
 
+// External streams
+// We need some temporary LeakyStreamInternals to which to stage information
+// about external streams when the user attempts to set them as current.
+static thread_local LeakyStreamInternals
```

Comment:
why does it have to be thread-local? Would the proper stream be passed to the backward thread? Probably we need a test for forward-backward on the external stream.

Reply:
It must be thread-local because it's the "temporary storage" of the current stream, which itself is thread-local. In other words: if two separate threads both try to set an external stream as current, they both must store it somewhere, but they must do so in two different places to avoid interfering with each other. I added a test that triggers precisely this scenario (and indeed that test fails if I remove `thread_local`).

Also, as I said, the "current stream" concept is already thread-local, and this PR doesn't modify that behavior. Thus I don't think anything will change with respect to autograd: it must already be able to handle this (i.e., I believe it retrieves the current stream during the forward and resets it during the backward) and thus it will work. If you really want me to add a test you'll need to tell me where because I don't know autograd.